bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java

```diff
@@ -622,91 +611,54 @@ void checkAllLedgers() throws BKAuditException, BKException,
-                                }
-                            }
+                admin.asyncOpenLedgerNoRecovery(ledgerId, (rc, lh, ctx) -> {
+                    if (Code.OK == rc) {
+                        checker.checkLedger(lh,
+                                // the ledger handle will be closed after checkLedger is done.
+                                new ProcessLostFragmentsCb(lh, callback),
+                                conf.getAuditorLedgerVerificationPercentage());
+                        // we collect the following stats to get a measure of the
+                        // distribution of a single ledger within the bk cluster
+                        // the higher the number of fragments/bookies, the more distributed it is
+                        numFragmentsPerLedger.registerSuccessfulValue(lh.getNumFragments());
+                        numBookiesPerLedger.registerSuccessfulValue(lh.getNumBookies());
+                        numLedgersChecked.inc();
+                    } else if (Code.NoSuchLedgerExistsException == rc) {
+                        if (LOG.isDebugEnabled()) {
+                            LOG.debug("Ledger {} was deleted before we could check it", ledgerId);
                         }
-                    });
-                }
+                        callback.processResult(Code.OK, null, null);
+                    } else {
+                        LOG.error("Couldn't open ledger {} to check : {}", ledgerId, BKException.getMessage(rc));
+                        callback.processResult(rc, null, null);
+                    }
+                }, null);
             };
 
             ledgerManager.asyncProcessLedgers(checkLedgersProcessor,
-                    new AsyncCallback.VoidCallback() {
-                        @Override
-                        public void processResult(int rc, String s, Object obj) {
-                            returnCode.set(rc);
-                            processDone.countDown();
-                        }
-                    }, null, BKException.Code.OK, BKException.Code.ReadException);
-            try {
-                processDone.await();
-            } catch (InterruptedException e) {
-                Thread.currentThread().interrupt();
-                throw new BKAuditException(
-                        "Exception while checking ledgers", e);
-            }
-            if (returnCode.get() != BKException.Code.OK) {
-                throw BKException.create(returnCode.get());
-            }
+                (rc, path, ctx) -> {
+                    if (Code.OK == rc) {
+                        FutureUtils.complete(processFuture, null);
+                    } else {
+                        FutureUtils.completeExceptionally(processFuture, BKException.create(rc));
+                    }
+                }, null, BKException.Code.OK, BKException.Code.ReadException);
+            FutureUtils.result(processFuture, BKException.HANDLER);
```

Comment:
If processing is done async, will get impacted by the below `admin.close()` and `client.close()`?

Reply:
processFuture is completed only when finished processing the last ledger, so checkAllLedgers is still sync, which is fine, since checkAllLedgers is executing in Auditor's own executor.